File: targetSum/targetSum.java
class Solution {
    public int findTargetSumWays(int[] nums, int target) {
        int sum = 0;
        for (int num : nums) sum += num;
        if (sum < target) return 0;
        int p = target + sum;
        if (p < 0 || p % 2 != 0) return 0;
        p = p / 2;
        
        int[][] dp = new int[nums.length + 1][p + 1];
        dp[0][0] = 1;
        for (int i = 1; i <= nums.length; i++) {
            for (int j = 0; j <= p; j++) {
                dp[i][j] += dp[i - 1][j];
                if (j >= nums[i - 1]) {
                    dp[i][j] += dp[i - 1][j - nums[i - 1]];
                }
            }
        }
        return dp[nums.length][p];
    }
}

File: 494.targetSum/targetSum.cpp
#include <iostream>
#include <vector>
#include <numeric>
using namespace std;

class Solution{
public:
  int findTargetSumWays(vector<int>& nums, int S) {
    int sum = accumulate(nums.begin(), nums.end(), 0);
    return sum < S || (S + sum) & 1 ? 0 : subsetSum(nums, (S + sum) >> 1);
  }
  int subsetSum(vector<int>& nums, int target) {
    vector<int> m(target+1, 0);
    m[0] = 1;
    for (int n : nums)
      for (int i = target; i >= n;i--)
	m[i] += m[i-n];
    return m[target];
  }
};

int main()
{
  Solution s;
  vector<int> vec(5, 1);
  cout << s.findTargetSumWays(vec, 3);
  cout << endl;
  return 0;
}
